audio-worklet/src/noisegate-script-processor.js

```diff
@@ -0,0 +1,217 @@
+   * @param {Float32Array} envelope array of amplitudes from envelope follower
+   * @return {Float32Array} weights numbers in the range 0 to 1 set in
+   *                                accordance with the threshold, the envelope,
+   *                                and attack and release. These will be
+   *                                multiplied by the corresponding
+   *                                value in the input.
+   */
+  computeWeights_(envelope) {
+    // When attack or release are 0, the weight changes between 0 and 1
+    // in one step.
+    let attackSteps = 1;
+    let releaseSteps = 1;
+    let attackLossPerStep = 1;
+    let releaseGainPerStep = 1;
+
+    // When attack or release are > 0, the associated weight changes between 0
+    // and 1 in the number of steps corresponding to the millisecond attack
+    // or release time parameters.
+    if (this.attack > 0) {
+      attackSteps = Math.ceil(this.context_.sampleRate * this.attack);
+      attackLossPerStep = 1 / attackSteps;
+    }
+    if (this.release > 0) {
+      releaseSteps = Math.ceil(this.context_.sampleRate * this.release);
+      releaseGainPerStep = 1 / releaseSteps;
+    }
+
+    // Compute an array of weights between 0 and 1 which will be multiplied with
+    // the channel depending on if the noise gate is open, attacking, releasing,
+    // or closed.
+    for (let i = 0; i < envelope.length; i++) {
+      // For sine waves, the envelope eventually reaches an average power of
+      // a^2 / 2. Sine waves are therefore scaled back to the original amplitude,
+      // but other waveforms or constant sources can only be approximated.
+      const scaledEnvelopeValue =
+          NoiseGate.toDecibel_(Math.sqrt(envelope[i]) * Math.sqrt(2));
+      if (scaledEnvelopeValue < this.threshold) {
+        const weight = this.previousWeight_ - attackLossPerStep;
+        this.weights_[i] = Math.max(weight, 0);
+      }
+      else {
+        const weight = this.previousWeight_ + releaseGainPerStep;
+        this.weights_[i] = Math.min(weight, 1);
+      }
+      this.previousWeight_ = this.weights_[i];
+    }
+    return this.weights_;
+  }
+
+  /**
+   * Computes the filter coefficent for the envelope filter.
+   * @param  {Number} timeConstant the time in seconds for filter to reach
+   *                               1 - 1/e of its value given a transition from
+   *                               0 to 1.
+   * @param  {Number} sampleRate the number of samples per second
+   * @return {Number} alpha weight governing envelope response
+   */
+  static getAlphaFromTimeConstant_(timeConstant, sampleRate) {
+    return Math.exp(-1 / (sampleRate * timeConstant));
+  }
+
+  /**
+   * Converts number into decibel measure.
+   * @param  {Number} linearValue the amplitude of the signal
+   * @return {Number} decibelValue the dBFS of the amplitude level
+   */
+  static toDecibel_(linearValue) {
+    return 20 * Math.log10(linearValue);
```

Comment:
I think technically this is wrong.  It should be 10*log10(x) because x should be a power term, not an amplitude term.

This might be ok if you're consistent with how you compute the linear value.